documentation/developers/mapping_instructions.md

```diff
@@ -0,0 +1,748 @@
+
+```json
+[
+  {
+    "name": "Person",
+    "type": "record",
+    "fields": [
+      {
+        "name": "id",
+        "type": "string",
+        "jsonldPredicate": "@id"
+      },
+      {
+        "name": "name",
+        "type": "string"
+      },
+      {
+        "name": "last_name",
+        "type": "string"
+      },
+      {
+        "name": "house",
+        "type": "string",
+        "jsonldPredicate": {
+          "_id": "eha.aether.jokes.House",
+          "_type": "@id"
+        }
+      }
+    ],
+    "namespace": "eha.aether.jokes"
+  },
+  {
+    "name": "House",
+    "type": "record",
+    "fields": [
+      {
+        "name": "id",
+        "type": "string",
+        "jsonldPredicate": "@id"
+      },
+      {
+        "name": "address",
+        "type": "string"
+      }
+    ],
+    "namespace": "eha.aether.jokes"
+  }
+]
+```
+
+</details>
+
+##### Steps
+
+Let's start with a suboptimal mappings
+
+<details open>
+    <summary><i>Mapping <b>(Hide/Show)</b></i></summary>
+
+```json
+[
+  [
+    "#!entity-reference#$.House[*].id",
+    "Person.house"
+  ],
+  [
+    "$.family_name",
+    "Person.last_name"
+  ],
+  [
+    "$.adults[*].name",
+    "Person.name"
+  ],
+  [
+    "$.children[*].name",
+    "Person.name"
+  ],
+  [
+    "$.address",
+    "House.address"
+  ],
+  [
+    "#!uuid",
+    "Person.id"
+  ],
+  [
+    "#!uuid",
+    "House.id"
+  ]
+]
+```
+
+</details>
+
+---
+<details>
+    <summary><i>First Pass Details <b>(Hide/Show)</b></i></summary>
+
+The first thing the extractor does is try to size the output requirements. In this case, based on the mappings and input, we can figure out that there will be 5 output entities; 4 `People` and 1 `Household`. We'll refer to the current state of the output at each mapping step.
+
+_Starting Objects_
+
+People
+```json
+[
+    {},
+    {},
+    {},
+    {}
+]
+```
+Households
+```json
+[
+    {}
+]
+```
+
+Instruction
+
+```json
+  [
+    "#!entity-reference#$.House[*].id",
+    "Person.house"
+  ]
+```
+_Resulting Objects_
+
+People
+```json
+[
+    {},
+    {},
+    {},
+    {}
+]
+```
+Households
+```json
+[
+    {}
+]
+```
+We don't have anything to reference in `$.House[*.id]` so this fails and is cached.
+
+Instruction
+
+```json
+  [
+    "$.family_name",
+    "Person.last_name"
+  ]
+```
+_Resulting Objects_
+
+People
+```json
+[
+    {"last_name": "Keaton"},
+    {"last_name": "Keaton"},
+    {"last_name": "Keaton"},
+    {"last_name": "Keaton"}
+]
+```
+Households
+```json
+[
+    {}
+]
+```
+This is a straght copy from a path that exists so it succeeds. Since there's only one value from `$.family_name` instead of four, the extractor guesses (correctly) that we should apply the same value to all Person objects. If this were just for the first Person, we could force the extractor to acknowledge this by adding the following instuction directly after: `["#!none", "Person.last_name"]`, which would set the `last_name` of everyone but the first (Buster) to null.
```

Comment:
```suggestion
This is a straight copy from a path that exists so it succeeds. Since there's only one value from `$.family_name` instead of four, the extractor guesses (correctly) that we should apply the same value to all Person objects. If this were just for the first Person, we could force the extractor to acknowledge this by adding the following instuction directly after: `["#!none", "Person.last_name"]`, which would set the `last_name` of everyone but the first (Buster) to null.
```